src/many-migration/src/lib.rs

```diff
@@ -362,117 +164,396 @@ impl<T, E> InnerMigration<'static, T, E> {
+        matches!(self.migration.r#type, MigrationType::Hotfix(_))
+    }
+
+    #[inline]
+    pub fn name(&self) -> &str {
+        self.migration.name()
+    }
+
+    #[inline]
+    pub fn description(&self) -> &str {
+        self.migration.description()
+    }
+
+    #[inline]
+    pub fn metadata(&self) -> &Metadata {
+        &self.metadata
+    }
+
+    #[inline]
+    pub fn disable(&mut self) {
+        self.enabled = false;
+    }
+
+    #[inline]
+    pub fn enable(&mut self) {
+        self.enabled = true;
+    }
+
+    #[inline]
+    pub fn is_enabled(&self) -> bool {
+        self.enabled
+    }
+
+    #[inline]
+    pub fn is_active(&self) -> bool {
+        self.active
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SingleMigrationConfig {
+    name: String,
 
     #[serde(flatten)]
     metadata: Metadata,
 }
 
-pub fn load_migrations<'a, 'b, E, T>(
-    registry: &'b [InnerMigration<'b, T, E>],
-    data: &'a str,
-) -> Result<BTreeMap<&'b str, Migration<'b, T, E>>, String> {
-    // TODO: Do not hardcode the deserializer
-    let config: Vec<MigrationJson> = serde_json::from_str(data).unwrap();
+impl<T, E> From<(&InnerMigration<T, E>, Metadata)> for SingleMigrationConfig {
+    fn from((migration, metadata): (&InnerMigration<T, E>, Metadata)) -> Self {
+        Self {
+            name: migration.name.to_string(),
+            metadata,
+        }
+    }
+}
+
+#[derive(Default, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct MigrationConfig(Vec<SingleMigrationConfig>);
 
-    // TODO: Cache this
-    // Build a BTreeMap from the linear registry
-    let registry = registry
-        .iter()
-        .map(|m| (m.name, m))
-        .collect::<BTreeMap<&'b str, &InnerMigration<'b, T, E>>>();
-
-    Ok(config
-        .into_iter()
-        .map(|io| {
-            let (&k, &v) = registry
-                .get_key_value(io.r#type)
-                .ok_or_else(|| format!("Unsupported migration type {}", io.r#type))?;
-            Ok((k, Migration::new(v, io.metadata, Status::Enabled)))
-        })
-        .collect::<Result<BTreeMap<_, _>, String>>()?
-        .into_iter()
-        .collect())
+impl MigrationConfig {
+    pub fn with_migration<T, E>(self, migration: &InnerMigration<T, E>) -> Self {
+        self.with_migration_opts(migration, Metadata::default())
+    }
+
+    pub fn with_migration_opts<T, E>(
+        mut self,
+        migration: &InnerMigration<T, E>,
+        metadata: Metadata,
+    ) -> Self {
+        self.0.push(SingleMigrationConfig {
+            name: migration.name.to_string(),
+            metadata,
+        });
+        self
+    }
+}
+
+impl<T: IntoIterator<Item = impl Into<SingleMigrationConfig>>> From<T> for MigrationConfig {
+    fn from(value: T) -> Self {
+        Self(value.into_iter().map(Into::into).collect())
+    }
+}
+
+pub struct MigrationSet<'a, T: 'a, E: 'a = many_error::ManyError> {
+    inner: BTreeMap<String, Migration<'a, T, E>>,
+    height: u64,
```

Comment:
Why are you keeping height here?

You don't use it anywhere but in `self.height = block_height;`

Reply:
Removed.